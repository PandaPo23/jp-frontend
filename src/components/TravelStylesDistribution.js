import React, { Fragment } from 'react';
import { withTranslation } from 'react-i18next';
import PropTypes from 'prop-types';

import { Grid, Text, RatingControl, DonutChart } from '../design-system';
import { numberString } from '../utils/propTypes';

export const TravelStylesDistribution = ({
  t,
  distribution,
  view,
  fontSize,
  width,
  ...pieChartProps
}) =>
  view === 'ratings' ? (
    <Grid
      width={width}
      gridTemplateColumns="2fr 1fr"
      gridColumnGap={1}
      gridRowGap={0}
    >
      {distribution.getData().map((ts, key) => (
        <Fragment key={key}>
          <Text uppercase fontSize={fontSize}>
            {t(ts.label, ts.name)}
          </Text>
          <RatingControl
            fractions={2}
            emptyIcon="empty-circle"
            fullIcon="circle"
            size={8}
            fullColor="secondary"
            readonly
            initialRating={ts.y / 10}
          />
        </Fragment>
      ))}
    </Grid>
  ) : (
    <DonutChart
      {...pieChartProps}
      data={distribution.getData()}
      colors={distribution.getColors()}
      t={t}
    />
  );

TravelStylesDistribution.propTypes = {
  t: PropTypes.func.isRequired,
  travelDistribution: PropTypes.object,
  view: PropTypes.oneOf(['pie', 'ratings']),
  fontSize: numberString,
  width: numberString,
};

export default withTranslation('common')(TravelStylesDistribution);
