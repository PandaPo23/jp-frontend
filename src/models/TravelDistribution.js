import { travelStyles } from './TravelStyle';
import { action, observable, decorate } from 'mobx';
import theme from '../design-system/theme';

const getColor = (style, key) => theme.colors.misc.branding.styles[style][key];
const normalize = (values) => {
  const sum = Object.values(values).reduce((s, v) => s + v, 0);
  return Object.keys(values).reduce((o, key) => {
    const v = values[key];
    if (v > 0) {
      o[key] = (100 * v) / sum;
    } else {
      o[key] = v;
    }
    return o;
  }, {});
};

export default class TravelDistribution {
  rawData = travelStyles.map((ts) => ({
    x: ts.abbr,
    y: 0,
    name: ts.name,
    onColor: getColor(ts.abbr, 'text'),
    color: getColor(ts.abbr, 'bg'),
  }));

  rawDataObject = {};

  constructor(values) {
    this.setData(values);
  }

  setData(values) {
    this.rawDataObject = values;
    const normalized = normalize(values);
    let dataCopy = this.rawData;
    Object.keys(normalize(values)).forEach((key) => {
      const value = normalized[key];
      const i = dataCopy.findIndex((d) => d.x === key);
      if (i >= 0) dataCopy[i].y = value;
    });
    this.rawData = dataCopy;
  }

  getData(includeZeroes = false) {
    return this.rawData.filter((d) => includeZeroes || d.y > 0);
  }

  getDataObject() {
    return this.rawDataObject;
  }

  getColors(includeZeroes = false) {
    return this.getData(includeZeroes).map(
      (_) => _.color || getColor(_.abbr, 'bg')
    );
  }

  async updateChartValues({ name, values }) {
    const dataObject = this.getDataObject();
    if (dataObject[name] !== values[0]) {
      dataObject[name] = values[0];
      await this.setData(dataObject);
    }
  }
}

decorate(TravelDistribution, {
  rawData: observable,
  rawDataObject: observable,
  setData: action,
  updateChartValues: action,
});
