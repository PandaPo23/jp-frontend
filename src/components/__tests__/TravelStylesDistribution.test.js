import { shallow } from 'enzyme';
import React from 'react';
import toJson from 'enzyme-to-json';

import { TravelStylesDistribution } from '../TravelStylesDistribution';
import TravelDistribution from '../../models/TravelDistribution';

const travelStyles = new TravelDistribution({
  oc: 20,
  an: 10,
  pp: 30,
  cu: 40,
  cn: 0,
});

describe('TravelStylesDistribution component', () => {
  it('renders piechart correctly', () => {
    const localProps = {
      distribution: travelStyles,
      view: 'ratings',
      t: (v) => v,
    };
    const wrapper = shallow(<TravelStylesDistribution {...localProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  // it('renders ratings correctly', () => {
  //   const localProps = {
  //     distribution: travelStyles,
  //     view: 'ratings',
  //     t: (v) => v,
  //   };
  //   const wrapper = shallow(<TravelStylesDistribution {...localProps} />);
  //   expect(toJson(wrapper)).toMatchSnapshot();
  // });
});
