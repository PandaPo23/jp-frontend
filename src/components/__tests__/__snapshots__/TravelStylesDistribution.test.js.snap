// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TravelStylesDistribution component renders piechart correctly 1`] = `
<Grid
  display="grid"
  gridColumnGap={1}
  gridRowGap={0}
  gridTemplateColumns="2fr 1fr"
>
  <Text
    uppercase={true}
  />
  <RatingControl
    color="border"
    direction="ltr"
    emptyIcon="empty-circle"
    emptySymbol={
      Object {
        "backgroundColor": "#ccc",
        "border": "5px double white",
        "borderRadius": "50%",
        "display": "inline-block",
        "height": 30,
        "width": 30,
      }
    }
    fractions={2}
    fullColor="secondary"
    fullIcon="circle"
    fullSymbol={
      Object {
        "backgroundColor": "black",
        "border": "5px double white",
        "borderRadius": "50%",
        "display": "inline-block",
        "height": 30,
        "width": 30,
      }
    }
    initialRating={2}
    onChange={[Function]}
    onClick={[Function]}
    onHover={[Function]}
    placeholderSymbol={
      Object {
        "backgroundColor": "red",
        "border": "5px double white",
        "borderRadius": "50%",
        "display": "inline-block",
        "height": 30,
        "width": 30,
      }
    }
    quiet={false}
    readonly={true}
    size={8}
    start={0}
    step={1}
    stop={5}
  />
  <Text
    uppercase={true}
  />
  <RatingControl
    color="border"
    direction="ltr"
    emptyIcon="empty-circle"
    emptySymbol={
      Object {
        "backgroundColor": "#ccc",
        "border": "5px double white",
        "borderRadius": "50%",
        "display": "inline-block",
        "height": 30,
        "width": 30,
      }
    }
    fractions={2}
    fullColor="secondary"
    fullIcon="circle"
    fullSymbol={
      Object {
        "backgroundColor": "black",
        "border": "5px double white",
        "borderRadius": "50%",
        "display": "inline-block",
        "height": 30,
        "width": 30,
      }
    }
    initialRating={1}
    onChange={[Function]}
    onClick={[Function]}
    onHover={[Function]}
    placeholderSymbol={
      Object {
        "backgroundColor": "red",
        "border": "5px double white",
        "borderRadius": "50%",
        "display": "inline-block",
        "height": 30,
        "width": 30,
      }
    }
    quiet={false}
    readonly={true}
    size={8}
    start={0}
    step={1}
    stop={5}
  />
  <Text
    uppercase={true}
  />
  <RatingControl
    color="border"
    direction="ltr"
    emptyIcon="empty-circle"
    emptySymbol={
      Object {
        "backgroundColor": "#ccc",
        "border": "5px double white",
        "borderRadius": "50%",
        "display": "inline-block",
        "height": 30,
        "width": 30,
      }
    }
    fractions={2}
    fullColor="secondary"
    fullIcon="circle"
    fullSymbol={
      Object {
        "backgroundColor": "black",
        "border": "5px double white",
        "borderRadius": "50%",
        "display": "inline-block",
        "height": 30,
        "width": 30,
      }
    }
    initialRating={4}
    onChange={[Function]}
    onClick={[Function]}
    onHover={[Function]}
    placeholderSymbol={
      Object {
        "backgroundColor": "red",
        "border": "5px double white",
        "borderRadius": "50%",
        "display": "inline-block",
        "height": 30,
        "width": 30,
      }
    }
    quiet={false}
    readonly={true}
    size={8}
    start={0}
    step={1}
    stop={5}
  />
  <Text
    uppercase={true}
  />
  <RatingControl
    color="border"
    direction="ltr"
    emptyIcon="empty-circle"
    emptySymbol={
      Object {
        "backgroundColor": "#ccc",
        "border": "5px double white",
        "borderRadius": "50%",
        "display": "inline-block",
        "height": 30,
        "width": 30,
      }
    }
    fractions={2}
    fullColor="secondary"
    fullIcon="circle"
    fullSymbol={
      Object {
        "backgroundColor": "black",
        "border": "5px double white",
        "borderRadius": "50%",
        "display": "inline-block",
        "height": 30,
        "width": 30,
      }
    }
    initialRating={3}
    onChange={[Function]}
    onClick={[Function]}
    onHover={[Function]}
    placeholderSymbol={
      Object {
        "backgroundColor": "red",
        "border": "5px double white",
        "borderRadius": "50%",
        "display": "inline-block",
        "height": 30,
        "width": 30,
      }
    }
    quiet={false}
    readonly={true}
    size={8}
    start={0}
    step={1}
    stop={5}
  />
</Grid>
`;
